### PR TITLE
GH-34918: [C++] Update vendored double-conversion 3.2.1

### LINKED ISSUE #34918: [C++] Update vendored double-conversion
### Describe the bug, including details regarding any error messages, version, and platform.

The current bundled double-conversion can't be built on LoongArch:

```text
-- Configuring incomplete, errors occurred!
See also "/home/loongson/tmp/aa/arrow/cpp/build/CMakeFiles/CMakeOutput.log".

loongson@loongson-pc:~/tmp/arrow/cpp/build$ make
In file included from /home/loongson/tmp/aa/arrow/cpp/src/arrow/vendored/double-conversion/double-conversion.h:31,
                 from /home/loongson/tmp/aa/arrow/cpp/src/arrow/util/double_conversion.h:20,
                 from /home/loongson/tmp/aa/arrow/cpp/src/arrow/util/formatting.h:35,
                 from /home/loongson/tmp/aa/arrow/cpp/src/arrow/pretty_print.cc:41:
/home/loongson/tmp/aa/arrow/cpp/src/arrow/vendored/double-conversion/utils.h:118:2: error: #error Target architecture was not detected as supported by Double-Conversion.
 #error Target architecture was not detected as supported by Double-Conversion.
  ^~~~~
make[2]: *** [src/arrow/CMakeFiles/arrow_objlib.dir/build.make:466：src/arrow/CMakeFiles/arrow_objlib.dir/pretty_print.cc.o] 错误 1
make[1]: *** [CMakeFiles/Makefile2:684：src/arrow/CMakeFiles/arrow_objlib.dir/all] 错误 2
make: *** [Makefile:141：all] 错误 2
```

### Component(s)

C++

### PR DESCRIPTION
**Pull request by liujiacheng777**
Modifications based on doubleConversion version 3.2.1
* Closes: #34918

### CODE REVIEW

**Comment by liujiacheng777** [comment]
https://github.com/apache/arrow/issues/34918#issue-1656542730

**Comment by kou** [comment]
Could you use pull request template content next time?

File: cpp/src/arrow/CMakeLists.txt [hunk 1]
```diff
@@ -237,12 +237,12 @@ set(ARROW_SRCS
...
     vendored/double-conversion/bignum.cc
-    vendored/double-conversion/double-conversion.cc
+    vendored/double-conversion/string-to-double.cc
+    vendored/double-conversion/double-to-string.cc
     vendored/double-conversion/bignum-dtoa.cc
     vendored/double-conversion/fast-dtoa.cc
     vendored/double-conversion/cached-powers.cc
     vendored/double-conversion/fixed-dtoa.cc
-    vendored/double-conversion/diy-fp.cc
     vendored/double-conversion/strtod.cc)
```
Comment (kou) [inline cpp/src/arrow/CMakeLists.txt:246]: Could you sort this list in alphabetical order?

Comment (liujiacheng777) [reply cpp/src/arrow/CMakeLists.txt]: Hello, I have sorted alphabetically in the new commit

**Comment by kou** [comment]
@anthonylouisbsb @projjal @praveenbingo You changed our vendored double-conversion in #9816. Is the change still needed?

Some Gandiva tests failed by upgrading vendored double-conversion (the change is revered): 

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/46724699#L2520

```text
[ RUN      ] TestGdvFnStubs.TestCastVARCHARFromFloat
C:/projects/arrow/cpp/src/gandiva/gdv_function_stubs_test.cc(400): error: Expected equality of these values:
  std::string(out_str, out_len)
    Which is: "1E-5"
  "1.0E-5"
[  FAILED  ] TestGdvFnStubs.TestCastVARCHARFromFloat (0 ms)
[ RUN      ] TestGdvFnStubs.TestCastVARCHARFromDouble
C:/projects/arrow/cpp/src/gandiva/gdv_function_stubs_test.cc(435): error: Expected equality of these values:
  std::string(out_str, out_len)
    Which is: "1E-5"
  "1.0E-5"
[  FAILED  ] TestGdvFnStubs.TestCastVARCHARFromDouble (0 ms)
```

If you really need the change, could you improve double-conversion in upstream instead of changing vendored double-conversion.

**Comment by ianmcook** [comment]
@liujiacheng777 As Kou says above, there were a few small changes made to the vendored double-conversion in #9816. It  https://github.com/apache/arrow/pull/9816/files#diff-d1cc5b70a5e980626bb70ae604a050d3393ac25a717a5a4c8dc40e8b5caf4b05
It looks like they were very small changes: less than 10 lines of code added, plus some comments. Could you please push a commit to this PR that ports these changes to the vendored double-conversion 3.2.1?

As Kou says above, we would like to upstream these changes, but porting the changes here will help, and I think it will fix the failing CI.

**Comment by kou** [comment]
FYI: #35135 added an update script. We need to update the patches under https://github.com/apache/arrow/tree/main/cpp/src/arrow/vendored/double-conversion/patches .

FYI: https://lists.apache.org/thread/mzb2qj4vkwhtbc64d1yfloc4s3r2po7j is the mailing list discussion about upstreaming.

**Comment by liujiacheng777** [comment]
`66% tests passed, 15 tests failed out of 44

Label Time Summary:
arrow-benchmarks    = 2663.27 sec*proc (44 tests)
benchmark           = 2663.27 sec*proc (44 tests)

Total Test time (real) = 980.08 sec

The following tests FAILED:
	  6 - arrow-compute-function-benchmark (Failed)
	  7 - arrow-compute-scalar-arithmetic-benchmark (Failed)
	  8 - arrow-compute-scalar-boolean-benchmark (Failed)
	 10 - arrow-compute-scalar-compare-benchmark (Failed)
	 11 - arrow-compute-scalar-if-else-benchmark (Failed)
	 12 - arrow-compute-scalar-random-benchmark (Failed)
	 13 - arrow-compute-scalar-round-benchmark (Failed)
	 14 - arrow-compute-scalar-set-lookup-benchmark (Failed)
	 15 - arrow-compute-scalar-string-benchmark (Failed)
	 16 - arrow-compute-scalar-temporal-benchmark (Failed)
	 18 - arrow-compute-vector-sort-benchmark (Failed)
	 19 - arrow-compute-vector-partition-benchmark (Failed)
	 20 - arrow-compute-vector-topk-benchmark (Failed)
	 21 - arrow-compute-vector-replace-benchmark (Failed)
	 22 - arrow-compute-vector-selection-benchmark (Failed)
Errors while running CTest`

@kou ，I tested using “cmake .. -DARROW_BUILD_BENCHMARKS=ON && ctest -j8 --output-on-failure” according to the official manual, and there were 15  errors in the test results. The reason for the error is similar to the function not being registered（Key error: No function registered with name: greater）. Do I need to turn on some test switches

**Comment by kou** [comment]
You may need to add `-DARROW_COMPUTE=ON`.

**Comment by liujiacheng777** [comment]
@kou  thankds，I turned on the DARROW_COMPUT and passed all the tests

`100% tests passed, 0 tests failed out of 44

Label Time Summary:
arrow-benchmarks    = 4639.70 sec*proc (44 tests)
benchmark           = 4639.70 sec*proc (44 tests)

Total Test time (real) = 691.67 sec`

besides，I integrated the first two commits：https://github.com/apache/arrow/pull/9816 && https://github.com/apache/arrow/pull/35135，upgrade double-conversion version from v3.1.5 to v3.2.1

**Comment by kou** [comment]
Thanks.
I started CI jobs.

If nobody reports our patches to upstream in this month, let's remove our patches and merge this.

**Comment by ianmcook** [comment]
I opened a PR to upstream the double-conversion change at https://github.com/google/double-conversion/pull/195

**Comment by kou** [comment]
@ursabot please benchmark

**Comment by ianmcook** [comment]
@kou do you want to merge this and then we can do https://github.com/apache/arrow/issues/35669 after?

**Comment by kou** [comment]
Yes.

But could you check the benchmark result before we merge this?

> ['Python', 'R'] benchmarks have high level of regressions. [ursa-i9-9960x](https://conbench.ursa.dev/compare/runs/32b4b885197b47e6be7afde1ef4eae14...ad7f8a6d1056462fa4f87b3a91e6f72b/)

It says that the tpch benchmarks were slow with this change.

(Should we re-run benchmark?)

**Comment by kou** [comment]
@ursabot please benchmark

**Comment by ianmcook** [comment]
@ursabot please benchmark

**Comment by ianmcook** [comment]
@ursabot please benchmark lang=R

**Comment by ianmcook** [comment]
It looks like we will have to wait for a while to re-run the benchmarks. It seems hard to tell whether the performance regressions are just noise.

**Comment by kou** [comment]
It seems that we can't run our benchmarks multiple times for one commit.

@liujiacheng777 Could you rebase on main to re-run our benchmarks? (Rebasing changes a commit ID.)

**Review by kou — APPROVED** [review]
+1

OK. I merge this.

**Pull request merged**